### ISSUE TITLE
Problems after Upgrade/fresh install of Piwigo

### ISSUE THREAD

**Issue by 0utl4w** [YunoHost-Apps/piwigo_ynh#146]
I updated my piwigo steup from 14.3 to 14.5. After the website doesn't show properly anymore. Tried a fresh install as well. Tried it via web-admin page and cli.

Console says the following:

Refused to apply style from 'https://gallery.test/home/yunohost.app/piwigo__2/_data/combined/1dfvcx1.css' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled.

Failed to load resource: the server responded with a status of 404 ()

Refused to execute script from 'https://gallery.test/home/yunohost.app/piwigo__2/_data/combined/4p4fz6.js' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.

Uncaught ReferenceError: CategoriesCache is not defined
    at admin.php?page=photos_add:265:25

- Hardware: Raspberry Pi at home
- YunoHost version: 12.0.7
- I have access to my server: Through SSH and through the webadmin
- Are you in a special context or did you perform some particular tweaking on your YunoHost instance?: no
- Using, or trying to install package version/branch: 14.5.0~ynh1
- If upgrading, current package version: 14.3.0~ynh1

### Expected behavior

App works and displays correctly after upgrade

**Comment by isAAAc** [comment]
Piwigo (piwigo) from 14.2.0~ynh1 to 14.5.0~ynh1

on VPS

```
root@gourzoot:~# yunohost --version
yunohost: 
  repo: stable
  version: 12.0.7
yunohost-admin: 
  repo: stable
  version: 12.0.4
yunohost-portal: 
  repo: stable
  version: 12.0.7
moulinette: 
  repo: stable
  version: 12.0.3
ssowat: 
  repo: stable
  version: 12.0.3
```



CMS is failing here too:

- on ffox, F12 console tells:
```
The resource from “https://gourzoot.org/yunohost/sso/?r=aHR0cHM6Ly9nb3Vyem9vdC5vcmcvaG9tZS95dW5vaG9zdC5hcHAvcGl3aWdvL19kYXRhL2NvbWJpbmVkL2h5ODY5cS5jc3M=” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff).
[piwigo](https://gourzoot.org/piwigo/)
The resource from “https://gourzoot.org/yunohost/sso/?r=aHR0cHM6Ly9nb3Vyem9vdC5vcmcvaG9tZS95dW5vaG9zdC5hcHAvcGl3aWdvL19kYXRhL2NvbWJpbmVkLzFmOGQ0ajAuanM=” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff).
[piwigo](https://gourzoot.org/piwigo/)
Loading failed for the <script> with source “https://gourzoot.org/home/yunohost.app/piwigo/_data/combined/1f8d4j0.js”. [piwigo:384:98](https://gourzoot.org/piwigo/)
Uncaught ReferenceError: jQuery is not defined
    <anonymous> https://gourzoot.org/piwigo/:387
[piwigo:387:9](https://gourzoot.org/piwigo/)
The resource from “https://gourzoot.org/yunohost/sso/?r=aHR0cHM6Ly9nb3Vyem9vdC5vcmcvaG9tZS95dW5vaG9zdC5hcHAvcGl3aWdvL19kYXRhL2NvbWJpbmVkL2lkbDhlZC5qcw==” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff).
[piwigo](https://gourzoot.org/piwigo/)
Loading failed for the <script> with source “https://gourzoot.org/home/yunohost.app/piwigo/_data/combined/idl8ed.js”. [piwigo:1:1](https://gourzoot.org/piwigo/)
```

- on chromium:
```
piwigo/:1 Refused to apply style from 'https://gourzoot.org/yunohost/sso/?r=aHR0cHM6Ly9nb3Vyem9vdC5vcmcvaG9tZS95dW5vaG9zdC5hcHAvcGl3aWdvL19kYXRhL2NvbWJpbmVkLzVkYzhvai5jc3M=' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled.
piwigo/:1 Refused to execute script from 'https://gourzoot.org/yunohost/sso/?r=aHR0cHM6Ly9nb3Vyem9vdC5vcmcvaG9tZS95dW5vaG9zdC5hcHAvcGl3aWdvL19kYXRhL2NvbWJpbmVkL2lkbDhlZC5qcw==' because its MIME type ('text/html') is not executable, and strict MIME type checking is enabled.
piwigo/:1 Refused to apply style from 'https://gourzoot.org/yunohost/sso/?r=aHR0cHM6Ly9nb3Vyem9vdC5vcmcvaG9tZS95dW5vaG9zdC5hcHAvcGl3aWdvL19kYXRhL2NvbWJpbmVkLzVkYzhvai5jc3M=' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled.
```

when i want to restore from backup:
> The following apps can't be restored because they are already installed: piwigo

restoration works _(and piwigo is working back)_ after removing the app first:
```bash
# yunohost app remove piwigo
# yunohost backup restore piwigo-pre-upgrade1
[...]
# Success! Restoration completed
```

thx for your help @Salamandar <3

**Comment by isAAAc** [comment]
`14.2.0~ynh1` to `15.1.0~ynh1`  same issue

**Comment by 0utl4w** [comment]
Still same issue with 15.1.0.

Found this online, sounds like a solution, but don't know what to do with it. :-P (sorry that it is in german, translate tool will help)

https://ekiwi.de/index.php/4978/die-resource-wurde-blockiert-wegen-eine-mime-typ-konfliktes-nosniff/

**Comment by ashledombos** [comment]
https://github.com/YunoHost-Apps/piwigo_ynh/pull/148 may solve the combined JS/CSS files error
and
https://github.com/YunoHost-Apps/piwigo_ynh/pull/149 may solve the access to images in /home/yunohost.app/piwigo/upload issue

**Comment by 0utl4w** [comment]
With the test-version 15.3.0 the problem is still the same

**Comment by ed0mm** [comment]
I can confirm this. I have piwigo on a site like foo.tld/piwigo, and I can't see anything, just broken interface, not even albums list (after logging in, which is like the only thing I'm able to do). I read somewhere it's different if your site is like piwigo.foo.tld.
Anyway I recently also saw that wallabag_ynh solved a long standing issue with login that had broke after a ynh update (see https://github.com/YunoHost-Apps/wallabag2_ynh/pull/229).
Also somebody on ynh forum could solve some things (almost everything?) on his piwigo by changing upload directories and so: https://forum.yunohost.org/t/piwigo-update-failed/32327/14

**Comment by Thatoo** [comment]
Same here trying to update from 14.0.0~ynh1 to 15.1.0~ynh1.

**Comment by lafartre** [comment]
I'm experiencing the same problem with stable (`15.1.0~ynh1`) and testing (`15.3.0~ynh1`) versions under Yunohost 12.0.11. 
The errors are different depending on the URL chosen for Piwigo.

If my Piwigo URL is `my_subdomain.my_domain.tld/piwigo_path` :
- the 1st connection to this URL redirects to `mydomain.mydomain.tld/piwigo_path/identification.php`, but I can't connect with my credentials, I get the following errors in my web browser (Firefox):
> Blocked : GET https://my_subdomain.my_domain.tld/yunohost/sso/?r=aHR0cHM6Ly9zZXJ2aWNlcy5zaG5hLmZyL2hvbWUveXVub2hvc3QuYXBwL3Bpd2lnby9fZGF0YS9jb21iaW5lZC91am1yaWUuY3Nz NS_ERROR_CORRUPTED_CONTENT
> Warning : Layout was forced before the page was fully loaded. If stylesheets are not yet loaded this may cause a flash of unstyled content. node.js:416:1
> Error : The resource from “https://my_subdomain.my_domain.tld/yunohost/sso/?r=aHR0cHM6Ly9zZXJ2aWNlcy5zaG5hLmZyL2hvbWUveXVub2hvc3QuYXBwL3Bpd2lnby9fZGF0YS9jb21iaW5lZC91am1yaWUuY3Nz” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff). identification.php

If my Piwigo URL is `piwigo_subdomain.my_subdomain.my_domain.tld/` :
- the 1st connection to this URL opens a login page where I can log in with my credentials
- the welcome page is displayed correctly ('_Hello <user>, your Piwigo photo gallery is empty!_' with the '_I want to add photos_' button)
- clicking on this button redirects me to the administration page `https://piwigo_subdomain.my_subdomain.my_domain.tld/admin.php?page=photos_add`, but the page style is not loaded correctly (no layout) and almost no action is possible (most of the menus are inaccessible, no hyperlink in particular for _Users_, _Tools_, _Configuration_, _Create a first album_), and I get the following errors in my web browser:

> Blocked : GET https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/852xq3.css NS_ERROR_CORRUPTED_CONTENT
> Blocked : GET https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/1sm6v5m.js NS_ERROR_CORRUPTED_CONTENT
> Error : The resource from “https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/852xq3.css” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff). admin.php
> Error : The resource from “https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/1sm6v5m.js” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff). admin.php
> Blocked : GET https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/1sm6v5m.js NS_ERROR_CORRUPTED_CONTENT
> Error : The resource from “https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/1sm6v5m.js” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff). admin.php
> Warning : Loading failed for the <script> with source “https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/1sm6v5m.js”. admin.php:412:98
> Error : Uncaught TypeError: jQuery(...).tipTip is not a function
>     <anonymous> https://piwigo_subdomain.my_subdomain.my_domain.tld/admin.php?page=photos_add:497
- and if I go to `https://piwigo_subdomain.my_subdomain.my_domain.tld/admin.php`, similar errors are returned:

> Blocked: GET https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/x1u274.css NS_ERROR_CORRUPTED_CONTENT
> Blocked: GET https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/1124bxm.js NS_ERROR_CORRUPTED_CONTENT
> Error : The resource from “https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/x1u274.css” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff). admin.php
> Error: The resource from “https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/1124bxm.js” was blocked due to MIME type (“text/html”) mismatch (X-Content-Type-Options: nosniff). admin.php
> Warning : Loading failed for the <script> with source “https://piwigo_subdomain.my_subdomain.my_domain.tld/home/yunohost.app/piwigo/_data/combined/1124bxm.js”. admin.php:330:98
> Error: Uncaught TypeError: jQuery(...).tipTip is not a function
>     <anonymous> https://piwigo_subdomain.my_subdomain.my_domain.tld/admin.php:432
> Error: Uncaught TypeError: jQuery(...).cluetip is not a function
>     <anonymous> https://piwigo_subdomain.my_subdomain.my_domain.tld/admin.php:345

All these css and js files are in 404 error, even though they actually exist.

Is it a problem of access permissions to these files, or of Piwigo or NGinx configuration?

**Comment by isAAAc** [comment]
same with `15.4.0~ynh1`

**Comment by moeenio** [comment]
> [#148](https://github.com/YunoHost-Apps/piwigo_ynh/pull/148) may solve the combined JS/CSS files error

it does not work on my side.

It seems that the `$conf['data_location']` is prepended to resource URLs on the web pages. `__DATA_DIR__` equals `/home/yunohost.app/piwigo/` thus a page tries to load `https://domain.tld/home/yunohost.app/piwigo/_data/...` regardless of the change in that PR.

Also the app [prepends its installation directory to `$conf['data_location']`](https://github.com/search?q=repo%3APiwigo%2FPiwigo%20%24conf%5B%27data_location%27%5D&type=code) when writing the data directory. With #148 it therefore creates  `/var/www/piwigo/home/yunohost.app/piwigo/_data/`. Seems to be why it currently uses a bunch of `..`.

I am not sure how to make the data directory stored in `/home/yunohost.app/piwigo/_data/` while having the web pages fetch `https://domain.tld/my_piwigo_path/_data/...`. That doesn't seem to be an option on the piwigo side.

Regardless, you'd then need an alias like in #149 to expose the _data directory properly, unless we store it in `/var/www/piwigo/_data` which I guess is not desirable.

Otherwise  maybe a symlink from `/var/www/piwigo/_data/` to `/home/yunohost.app/piwigo/_data/` would do the job and require no nginx alias.

**Comment by 0utl4w** [comment]
I found a solution here: https://forum.yunohost.org/t/piwigo-update-failed/32327/14 

It the last post from leosw.

**Comment by moeenio** [comment]
TL;DR : that solution uses symlinks

**Comment by 0utl4w** [comment]
> TL;DR : that solution uses symlinks

true, it only works for a fresh install but not for an upgrade.

**Comment by moeenio** [comment]
> TL;DR : that solution uses symlinks

would that be accepted for inclusion as part of the install/upgrade scripts? @yalh76

**Comment by yalh76** [comment]
well, the issue comes from https://github.com/YunoHost-Apps/piwigo_ynh/pull/129 since 14.5.0~ynh1

**Comment by moeenio** [comment]
I don't see anything related in that pr so you mean this is an issue upstream with the new version of piwigo? did you reach out to that?

**Comment by yalh76** [comment]
> I don't see anything related in that pr so you mean this is an issue upstream with the new version of piwigo? did you reach out to that?

After more search, it comes from #124 when `$conf['data_location'] = '__DATA_DIR__/_data/';` include in `14.3.0~ynh1`

**Comment by isAAAc** [comment]
it works <3

thx you @yalh76  ㊗
